### ISSUE TITLE
Yet more QA on missing data

### ISSUE THREAD

**Issue by danrademacher** [GreenInfo-Network/nyc-crash-mapper-etl-script#16]
Christine says we're missing data. #15 should resolve a bunch of that, but she also sent this CSV and wants some data QA: https://docs.google.com/spreadsheets/d/1T9-ENN7uuhHabm3kYhl7XulO8th5qYgbZA2eaSNrOnE/edit#gid=1551094594

I responded asking her what the query was that generated this doc. I can't divine it from the data itself -- seems like random days grouped by borough.

**Comment by danrademacher** [comment]
From client:

Here are two data sets  

* NYPD_Motor_Vehicle_Collisions-FATALITIES.csv
fatalities citywide from January 1st, 2018 to November 13, 2018

* NYPD_Motor_Vehicle_Collisions-INJURIES.csv
Injuries citywide from January 1st 2018. to November 13, 2018 

[20181114_crashmapperQA.zip](https://github.com/GreenInfo-Network/nyc-crash-mapper-etl-script/files/2583277/20181114_crashmapperQA.zip)

Here is the the source in case you want to make subsets to pinpoint the problems. 
https://data.cityofnewyork.us/Public-Safety/NYPD-Motor-Vehicle-Collisions/h9gi-nx95/data

I amk not certain we even have a problem. Going to ask her now that the missing data is restored.

**Comment by danrademacher** [comment]
OK, I repeated past QA processed and results are here:L
https://docs.google.com/spreadsheets/d/1q3yi39yAEYPT-vc2NMWUCasDUUI2ASbe0azD6q_ZCEY/edit#gid=187881358

We are missing more than 3K crashes. I have isolated nearly all to May and June of this year.

@gregallensworth I reminded myself that https://github.com/GreenInfo-Network/nyc-crash-mapper-etl-script/tree/master/backlog exists. Is that something we could rerun on 2018 YTD and maybe pick up these missing crashes? Or is it supposedly already running and we have a new source of missing crashes?

If the former, let's try it. If the latter, I want to report back to Christine and assess level of effort vs resources.

**Comment by gregallensworth** [comment]
I've run the `backlog.py` script, it only discovered *several* missing crashes, not a significant chunk of what you're reporting. I also checked over the main ETL script's `find_updated_killcounts()` which looks for later-updated records and ran it patched to look back to February; it too found only *several* variances, not a significant proportion of what you are reporting. I then ran some modified examinations of the ETL script for the period of *2018-05 thru 2018-06* and came up with nothing new.

During my examination though, I did find some crashes which had a `date_entered` in November 2018 but a `date` of April 2016. While not directly relevant to May 2018 right now, it is disturbing to note that come 2020 if we were to run SODA/CARTO variances for 2018, we may find new "missing" records where NYC waited 2 years to enter them.

To attack the current variance, though, I'll need to move on to the tedious step downloading a dump of every single crash in *2018-05* from CARTO and from SODA, then comparing them, to see what I can track down.

**Comment by gregallensworth** [comment]
### ETL Results, Raw Comparison

I fetched raw lists of crashes from SODA and CARTO (see fixtallies documentation for queries used) in CSV format, then removed and reorganized columns in the SODA CSVs to match those of CARTO CSVs.

Summary counts loaded in LibreOffice are as follows *and seem to align well within acceptable variances*. Even the -5 injuries, could well be a bus crash with injuries reported months later.

```
201805           SODA       CARTO     DIFF
Crash Count     20720       20721     +1
Total Inj        5647        5642     -5
Total Kil          13          13     OK

201806           SODA       CARTO      DIFF
Crash Count     20672       20670      +2
Total Inj        5590        5591      +1
Total Kil          21          21      OK
```

**Bottom line here, is that the ETL script seems to be doing quite well** with CARTO varying well under 1% from the NYC Socrata data.

**Comment by gregallensworth** [comment]
### CrashMapper (Map View) Summary Counts

Expanding on https://github.com/GreenInfo-Network/nyc-crash-mapper-etl-script/issues/16#issuecomment-442951923

CrashMapper's own summary table at the bottom gives very different results than those represented in its own data exported above. If we were to go by the CrashMaper summary table at the bottom (manually adding the injury/kill crashes with harmless crashes, as the app won't display both at once) we find that *CrashMapper reports 3002 fewer crashes* over this two-month period, as well as significantly lower-than-expected injury and fatality counts. *This is the CrashMapper summary table tallying data in CARTO, therefore a behavior of CrashMapper and not of the ETL script.*

```
201805 through 201806
Crashes Harmful 5708 + Harmless 33239 = 38947    expected 41391 / underreporting 2444
Injuries                                 7861    expected 11233 / underreporting 3372
Fatalities                                 28    expected 34    / underreporting 6
```
I ran the SQL queries manually, as well as a few variants e.g. to include both injury/kill and harmless crashes at once, as the UI wil not show both at once. I found something very interesting in regards to the `persons_injured` and `persons_killed` tallies. *The `SUM()` aggregation somehow causes these records to be omitted.*


```
-- SQL queries for 201805 thru 201806, as performed and displayed in the CrashMapper summary table
-- separate UI and query happens for harmless crashes vs injury/kill crashes, so two separate queries BUT here I merged into one by dropping the "and injury > 0" clause
-- not surprisingly, the results here are what's displayed in the CrashMapper

SELECT
    COUNT(c.cartodb_id) as total_crashes,
    SUM(c.number_of_pedestrian_injured + c.number_of_cyclist_injured + c.number_of_motorist_injured) as persons_injured,
    SUM(c.number_of_pedestrian_killed + c.number_of_cyclist_killed + c.number_of_motorist_killed) as persons_killed
FROM crashes_all_prod c
WHERE
    ( year::text || LPAD(month::text, 2, '0') <= '2018' || LPAD(6::text, 2, '0') ) AND ( year::text || LPAD(month::text, 2, '0') >= '2018' || LPAD(5::text, 2, '0') )

-- however, this is weird...
-- if we go by the number_of_persons_injured and number_of_persons_killed tally instead
-- then our numbers match perfectly; even the crash count (row count) matches

SELECT
    COUNT(c.cartodb_id) as total_crashes,
    SUM(c.number_of_persons_injured) as persons_injured,
    SUM(c.number_of_persons_killed) as persons_killed
FROM crashes_all_prod c
WHERE
    ( year::text || LPAD(month::text, 2, '0') <= '2018' || LPAD(6::text, 2, '0') ) AND ( year::text || LPAD(month::text, 2, '0') >= '2018' || LPAD(5::text, 2, '0') )
```

I checked around and determined that there are no rows with a NULL value for any of the `number_of_X_injured` and `number_of_X_killed` fields, so I'm not sure what's going on.

*At this point the most expedient fix is probably just to fix the query, so it uses `SUM(c.number_of_persons_injured)` AND `SUM(c.number_of_persons_killed)` so the right numbers come out.*

**Comment by gregallensworth** [comment]
### Front-end fixes: Map view

The stats displayed now agree with the tabular analysis. Below are counts for injury/fatality crashes and harmless crashes: 34 fatality, 11233 injury, total 41391 crashes

![image](https://user-images.githubusercontent.com/3117633/49250423-a368bf00-f3d3-11e8-80b2-b78398cee575.png)

![image](https://user-images.githubusercontent.com/3117633/49250455-b8dde900-f3d3-11e8-82fe-d406dc5fcb89.png)

**Comment by gregallensworth** [comment]
### Details review

Thumbing passively through June 2018, as a general rule I am seeing a lot of persons injured but with 0/0/0 for the cyc/ped/mot counts (no actual tallies as that would take time, but a general impression by sorting the rows and skipping to those where persons injured >0 and the other 3 are all 0). As such, the `persons_Injured` (which may be what NYC reports in their dashboard) would be way off from the actual sum of the 3 injury "roles".

Crash ID 3912235 in June 2018
2 persons injured, but 0/0/0 cyc/mot/ped
later revised to 2 motorist

Crash ID 3913343 in June 2018
3 persons injured, but 0/0/0 cyc/mot/ped
later revised to 3 motorist

Crash ID 3912547 in June 2018
2 persons injured, but 0/0/0 cyc/mot/ped
later revised to 1 motorist (still not 2 total!)

Crash ID 3912770 in June 2018
3 persons injured, 0/0/0 details
later revised to 3 motorist

Crash ID 3913183 in June 2018
3 persons, 0/0/0 detail
later revised to 3 pedestrian injuries

And so on and so on. That was the first few hundred rows... Taken 1-2 variants at a time over maybe 5,000 injurious crashes, this would certainly account for what we are seeing. *And if NYC/Socarata dashboards are based on the total persons fields, instead of the sum of cyc+mot+ped, then the NYC/Socarata data will over-report relative to a more detailed "by role" view.*

**Comment by gregallensworth** [comment]
### Fixtallies to Fix Tallies

I fired up the `fixtallies` script, and fed it full exports of CARTO and SODA dated *2015-01-01 through 2018-10-01* The difference is *3057* crashes in which the tallies are different in CARTO than they now appear in SODA. **After running this, the numbers for May + June 2018 match much more closely.**

![image](https://user-images.githubusercontent.com/3117633/49303519-b5536c00-f47e-11e8-91a3-2aeaa72243d8.png)

**Fatalities for this 2-month period:** 34, **exactly matching** SODA

**Injuries for this 2-month period:** 11,170 versus 11,237 in SODA. Diff is 67, a **variance of 0.5%**

Still, I am bothered that there would be a variance of more than 1 or 2 if records were updated in NYC/SODA overnight or early this morning (3 hour time zone diff). So I do want to find out that variance of 67 injuries in this two-month period.
